samcli/lib/test_runner/test_runner_template_generator.py

```diff
@@ -0,0 +1,347 @@
+        return boto3.client("resourcegroupstaggingapi").get_resources(tag_filters=tag_filters)
+    except Exception as query_error:
+        LOG.error("Failed to call get_resources from resource group tagging api: %s", query_error)
+        return None
+
+
+def __generate_statement_list(tag_filters: dict) -> List[dict]:
+    """
+    Generates a list of IAM statements with default action permissions for resources with the given tags.
+
+    Parameters
+    ----------
+    tag_filters : dict
+        The tag filters to restrict output to only those resources with the given tags.
+
+        NOTE: https://docs.aws.amazon.com/cli/latest/reference/resourcegroupstaggingapi/get-resources.html#options
+
+    Returns
+    -------
+    List[dict]
+        The list of JSON objects representing IAM statements corresponding to the resources specified by the given tags.
+    """
+
+    tagging_api_response = __query_tagging_api(tag_filters)
+
+    if tagging_api_response is None:
+        LOG.error("Failed to receive get_resources response, cannot generate any IAM statements.")
+        return []
+
+    iam_statements = []
+
+    for resource in tagging_api_response["ResourceTagMappingList"]:
+        new_statement = {
+            "Effect": "Allow",
+            "Action": [],
+            "Resource": "",
+        }
+        arn = resource["ResourceARN"]
+
+        new_statement["Action"].extend(
+            # Default permissions are commented out
+            ["# " + action for action in __get_permissions(arn)]
+        )
+
+        # https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-control-access-using-iam-policies-to-invoke-api.html
+        if "# execute-api:Invoke" in new_statement["Action"]:
+            apiId = __extract_api_id_from_arn(arn)
+            new_statement[
+                "Resource"
+            ] = f"!Sub arn:${{AWS::Partition}}:execute-api:${{AWS::Region}}:${{AWS::AccountId}}:{apiId}/<STAGE>/GET/<RESOURCE_PATH>"
+        else:
+            new_statement["Resource"] = arn
+
+        iam_statements.append(new_statement)
+    return iam_statements
+
+
+def __comment_out_default_actions(raw_yaml_string: str) -> str:
+    """
+    Formats the given YAML string to remove quotes and move `#` to the other side of the list item `-`.
+
+    This is to allow default permissions to actually be commented out.
+    This string replacement will produce deterministic results, since the `raw_yaml_string` is deterministic.
+    There will be no other instances of `- # ` other than the ones created during the template generation process.
+
+    The purpose is to establish a barrier of consent, to prevent customers from accidentally granting permissions that they did not mean to.
+
+    e.g.
+
+        `- '# lambda:InvokeFunction'` => `# - lambda:InvokeFunction`
+    """
+    # Move comment hashtag to be in front of list items so they're properly commented out
+    return raw_yaml_string.replace("'", "").replace("- # ", "# - ")
```

Comment:
nit: I wonder if it would be cleaner to have the base template as YAML with a template variable for the IAM policy, then inject it with `"\n".join("# - " + action for action in actions)`. At a glance it would get rid of the need for `cfn-flip` and this hack

Reply:
Update: Found a way to do this. We can keep everything in YAML, avoid the `cfn-flip` dependency, and the hack.